src/middlewared/middlewared/plugins/reporting/netdata_web.py

```diff
@@ -0,0 +1,83 @@
+import random
+import os
+import string
+import shutil
+import time
+
+from middlewared.schema import accepts, returns, Password
+from middlewared.service import cli_private, job, pass_app, periodic, private, CallError, Service
+from middlewared.utils import MIDDLEWARE_RUN_DIR
+from passlib.apache import HtpasswdFile
+
+
+BASIC_FILE = f'{MIDDLEWARE_RUN_DIR}/netdata-basic'
+
+
+class ReportingService(Service):
+
+    @private
+    async def netdataweb_basic_file(self):
+        return BASIC_FILE
+
+    @cli_private
+    @accepts()
+    @returns(Password('password'))
+    @pass_app()
+    def netdataweb_generate_password(self, app):
+        """
+        Generate a password to access netdata web.
+        That password will be stored in htpasswd format for HTTP Basic access.
+        """
+        if app.authenticated_credentials.is_user_session:
+            authenticated_user = app.authenticated_credentials.user['username']
+        else:
+            raise CallError('This method needs to be called from an authenticated user only.')
+
+        if not os.path.exists(BASIC_FILE):
+            with open(os.open(BASIC_FILE, flags=os.O_CREAT, mode=0o640)):
+                shutil.chown(BASIC_FILE, 'root', 'www-data')
+
+        ht = HtpasswdFile(BASIC_FILE, autosave=True, default_scheme='bcrypt')
+        password = ''.join(random.choice(
```

Comment:
I've created a helper function that does this. It's in utils.crypto

Reply:
Sweet, changed